### PR TITLE
 chore(licenseinfo): Reporting Improvements Part 2 #433 

### PR DESCRIPTION
**Pull request by dreh23**
- Development Detail table
  - Additional Requirement table

Signed-off-by: Johannes Amorosa <johannes.amorosa@endocode.com>

### CODE REVIEW

File: libraries/lib-datahandler/src/main/thrift/licenseinfo.thrift [hunk 1]
```diff
@@ -70,6 +70,10 @@ struct LicenseInfoParsingResult {
     31: optional string name,
     32: optional string version,
     33: optional string componentType,
+
+    42: optional set<string> componentLanguages,
```
Comment (mcjaeger) [inline libraries/lib-datahandler/src/main/thrift/licenseinfo.thrift]: maybe the point here, agreeing with @maierthomas is that we have this not as part of the backend service API for the license info service - it appears rather as information that is processed internally to generate the document. Thus I also propose to pull this off from the API of the license inform service.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/LicenseInfoHandler.java [hunk 2]
```diff
@@ -381,6 +382,24 @@ protected LicenseInfoParsingResult noSourceParsingResult(String message) {
         parsingResults.forEach(result -> {
             if( component != null) {
                 result.setComponentType(toString(component.getComponentType()));
+
+                if(component.isSetLanguages()) {
+                    result.setComponentLanguages(component.getLanguages());
+                } else {
+                    result.setComponentLanguages(Stream.of("Unknown languages.").collect(Collectors.toSet()));
```
[the remark this answers is not in the document]
Comment (dreh23) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/LicenseInfoHandler.java]: Done

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/LicenseInfoHandler.java [hunk 3]
```diff
@@ -42,6 +42,7 @@
 import java.util.Map.Entry;
```
[the remark this answers is not in the document]
Comment (dreh23) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/LicenseInfoHandler.java]: Done

**Comment by mcjaeger** [comment]
tested, works, but is conflicting
(also there seem to be a problem with SPDX files)

**Comment by mcjaeger** [comment]
@dreh23 I wondered in testing where is the setting for platform in the UI - was not able to test it

**Comment by mcjaeger** [comment]
@dreh23 seems to be conflicting, maybe because of the merge of the previous pull request PR#438

**Comment by mcjaeger** [comment]
@dreh23 please try to resolve the conflicts

**Comment by mcjaeger** [comment]
@dreh23 signed commit?

**Comment by imaykay** [comment]
Do we have any issue ticket that explains what the feature idea behind this PR is? I mean, tests seem to be already successful, but for code review it also interesting what should be achieved - at least for me.

/cc @mcjaeger @dreh23 @hemarkus

**Comment by dreh23** [comment]
@imaykay There are tickets https://github.com/siemens/sw360/projects/1 and there is a reference documentation that was provided which describes some aspects of the features. If @mcjaeger agrees I can share this doc with you.

On a side note I am not any more involved in the development of sw360. Please contact @hemarkus  for any further questions.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 4]
```diff
@@ -151,6 +157,8 @@ private void fillReportDocument(
             String projectDescription = project.getDescription();
 
             fillOverview3rdPartyComponentTable(document, projectLicenseInfoResults);
+            fillDevelopmentDetailsTable(document, project, user);
```
[the remark this answers is not in the document]
Comment (hemarkus) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: done

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 5]
```diff
@@ -366,6 +378,87 @@ private void writeComponentSubsections(XWPFDocument document, Collection<License
...
+
+            String operatingSystems = component.getOperatingSystems().isEmpty() ? "Unknown operating systems" : String.join(" ", component.getOperatingSystems());
+            row.addNewTableCell().setText(operatingSystems);
+            LOGGER.info(String.format("oses: %s", operatingSystems));
+
+            String langs = component.getLanguages().isEmpty() ? "Unknown languages" : String.join(" ", component.getLanguages());
+            row.addNewTableCell().setText(langs);
+            LOGGER.info(String.format("langs: %s", langs));
+
+            String platforms = component.getSoftwarePlatforms().isEmpty() ? "Unknown platforms" : String.join(" ", component.getSoftwarePlatforms());
```
[the remark this answers is not in the document]
Comment (hemarkus) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: Even though I could not reproduce this I see the chance to hit a null reference due to Thirft. I added a check before accessing it.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 6]
```diff
@@ -169,10 +177,14 @@ private void fillReportDocument(
 
             fillSpecialOSSRisksTable(document, project, obligationResults);
 
+            fillReleaseBulletList(document, projectLicenseInfoResults);
+            fillReleaseDetailList(document, projectLicenseInfoResults, includeObligations);
+            fillAdditionalRequirementsTable(document, obligationResults);
```
[the remark this answers is not in the document]
Comment (hemarkus) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: Actually those where not even supposed to be called here. Removed.

File: backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java [hunk 7]
```diff
@@ -366,6 +378,87 @@ private void writeComponentSubsections(XWPFDocument document, Collection<License
...
+        XWPFTable table = document.getTables().get(4);
+        int currentRow = 1;
+        for( ObligationKey key :  collatedObligations.keySet()) {
+            ArrayList<String> licenses = collatedObligations.get(key);
+            XWPFTableRow row = table.insertNewTableRow(currentRow++);
+            String licensesString = String.join(" ", licenses);
+            row.addNewTableCell().setText(key.topic);
+            row.addNewTableCell().setText(licensesString);
+            row.addNewTableCell().setText(key.text);
+        }
```
[the remark this answers is not in the document]
Comment (hemarkus) [reply backend/src/src-licenseinfo/src/main/java/org/eclipse/sw360/licenseinfo/outputGenerators/DocxGenerator.java]: See above

**Comment by imaykay** [comment]
Code looks better now, but the latest changes do change a few things in the generated document. So maybe @mcjaeger, you want to retest if the result is still what you want it to be.

@hemarkus it would be nice if you could squash your commits into a single one like it is mentioned here https://github.com/eclipse/sw360/wiki/Dev-DoD-and-Style in section "Commit style".

**Pull request merged**